README.md
# AlatPay Payment SDK

The AlatPayFlutter class is a flutter widget that allows you to integrate WEMA Bank's Alatpay payment gateway into your flutter project. It provides a WebView that loads an Alatpay payment page with the transaction details to be processed. The class handles the response from the payment gateway and provides callback functions for handling successful transactions and errors. The class requires parameters such as metaData, businessId, apiKey, and amount to process the payment. It supports Android, iOS and Web platforms. Other platforms coming soon.

## Author

[![Linkedin Badge](https://img.shields.io/badge/-Gideon-black?style=for-the-badge&logo=Linkedin&logoColor=white&link=https://www.linkedin.com/in/gideonchukwuoma/)](https://www.linkedin.com/in/gideonchukwuoma/)
[![Twitter Badge](https://img.shields.io/badge/-@egideonchuks-1ca0f1?style=for-the-badge&logo=twitter&logoColor=white&link=https://twitter.com/egideonchuks)](https://mobile.twitter.com/egideonchuks)
[![My Badge](https://img.shields.io/badge/-Gideon.Chukwuoma.Chimemerie-green?style=for-the-badge&logo=Google-Chrome&logoColor=white&link=https://gideonchukwuoma.vercel.app/)](https://gideonchukwuoma.vercel.app/)

## Features

* Integration with Alatpay Payment Gateway: The AlatPayFlutter class allows you to easily integrate the Moamalat payment gateway into your Flutter app.

* WebView Integration: The class provides a WebView that loads the Alatpay's payment page with the transaction details to be processed.

* Transaction Handling: The class handles the response from the payment gateway and provides callback functions for handling successful transactions and errors.

* Secure Transactions: The class generates a secure hash required by alatpay to ensure secure transactions.

* Customizable: The class is customizable and allows you to configure parameters such as apiKey, businessId, currency, amount, and metaData to process the payment.

Overall, the AlatPayFlutter class provides a convenient and secure way to integrate the AlatPay payment gateway into your Flutter app

## Documentation

`businessId` : The business ID obtained from AlatPay.

`amount` : The amount to be charged for the transaction.

`apiKey` : The api key obtained from AlatPay.

`onTransaction` : A callback function that is called when the payment is completed successfully.

`onClose`: A callback function that is called when the service is closed.

<!-- `alatpay_flutter:` AlatPayFlutter dependency.

`AlatPayFlutter`: The main AlatPayFlutter widget that handles the whole process. -->

## Installation

To use this package, add `alatpay_flutter` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

```yaml {"id":"01HJZ0C4E8NYR3TDB3TYMVYFWD"}
dependencies:
  alatpay_flutter: latest version
```

## Example

How to use the alatpay_flutter package.

```dart {"id":"01HJZ0C4E8NYR3TDB3V1EYNWDG"}

import 'dart:io';

import 'package:alatpay_flutter/alatpay_flutter.dart';
import 'package:flutter/material.dart';
import 'package:get/route_manager.dart';

import 'screens/success.dart';
import 'src/constants.dart';
import 'src/controllers.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'AlatPay',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: kAccentColor),
        useMaterial3: true,
      ),
      home: const PaymentScreen(),
    );
  }
}

class PaymentScreen extends StatefulWidget {
  const PaymentScreen({Key? key}) : super(key: key);

  @override
  State<PaymentScreen> createState() => _PaymentScreenState();
}

class _PaymentScreenState extends State<PaymentScreen> {
//============= Boolean variables ===============\\
  bool isLoading = false;

//============= Function ===============\\
  void pay() {
    String apiKey = "alatPayPrimaryKey";
    String businessId = "alatPayBuinessId";
    String email = "johndoe@gmail.com";
    String phone = "+2348047589684";
    String firstName = "John";
    String lastName = "Doe";
    String currency = 'NGN';
    String amount = "20000";

    setState(() {
      isLoading = true;
    });

    //=========== Not so important ===========\\

    // Map meta = {
    //   "the_order_id": "order_id",
    //   'client_id': "user_id"
    // };

    //=========== End =============\\
    try {
      Navigator.push(
        context,
        MaterialPageRoute(builder: (context) {
          return AlatPayFlutter(
            apiKey: apiKey,
            businessId: businessId,
            email: email,
            phone: phone,
            firstName: firstName,
            lastName: lastName,
            currency: currency,
            amount: amount,
            // metaData: meta,
            onClose: () {
              Get.close(2);
            },
            onTransaction: (response) async {
              //============ Do something =============\\
              Get.to(const SuccessScreen());
            },
          );
        }),
      );
      setState(() {
        isLoading = false;
      });
    } on SocketException {
      ApiProcessorController.errorSnack("Please connect to the internet");
    } catch (e) {
      consoleLog(e.toString());
    }
    setState(() {
      isLoading = false;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        elevation: 10,
        title: Text(
          "AlatPay",
          style: TextStyle(
            color: Theme.of(context).appBarTheme.backgroundColor,
            fontSize: 20,
            fontWeight: FontWeight.w600,
          ),
        ),
      ),
      body: SafeArea(
          child: Center(
        child: ElevatedButton(
          onPressed: pay,
          style: ElevatedButton.styleFrom(
            disabledBackgroundColor: Colors.grey.shade300,
            backgroundColor: Theme.of(context).primaryColor,
            elevation: 10,
            shape: RoundedRectangleBorder(
              borderRadius: BorderRadius.circular(20),
            ),
            shadowColor: Colors.black.withOpacity(0.4),
          ),
          child: isLoading
              ? const Padding(
                  padding: EdgeInsets.all(10),
                  child: CircularProgressIndicator(color: Colors.white),
                )
              : const Text(
                  "Pay now",
                  textAlign: TextAlign.center,
                  style: TextStyle(
                    color: Colors.white,
                    fontSize: 18,
                    fontFamily: "Sen",
                    fontWeight: FontWeight.w700,
                  ),
                ),
        ),
      )),
    );
  }
}


```


CHANGELOG.md
## 0.1.0

* Updated pubspec.yaml
* Updated README.md

## 0.0.9

* Updated pubspec.yaml
* Updated README.md

## 0.0.8

* Updated pubspec.yaml
* Updated README.md

## 0.0.7

* Added routing to a success screen.
* Removed unnecessary files and folders.
* Updated README.md

## 0.0.6

* Moved the payment screen to the main.dart file.

## 0.0.5

* Added support for all platforms.

## 0.0.4

* Added `getx` and `font_awesome_flutter` dependencies.

## 0.0.3

* Updated the UI.

## 0.0.2

* Updated the README.md file.

## 0.0.1

* Initial version.
* Added License.txt and other necessary files.


lib/alatpay_flutter.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:webviewx/webviewx.dart';

class AlatPayFlutter extends StatefulWidget {
  final String apiKey;
  final String businessId;
  final Map? metaData;
  final String email;
  final String phone;
  final String firstName;
  final String lastName;
  final String currency;
  final String amount;
  final Function(dynamic response) onTransaction;
  final Function()? onClose;
  const AlatPayFlutter({
    super.key,
    required this.apiKey,
    required this.businessId,
    this.metaData,
    required this.email,
    this.phone = '',
    this.firstName = '',
    this.lastName = '',
    this.currency = 'NGN',
    required this.amount,
    required this.onTransaction,
    this.onClose,
  });

  @override
  AlatPayFlutterState createState() => AlatPayFlutterState();
}

class AlatPayFlutterState extends State<AlatPayFlutter> {
  late WebViewXController webviewController;
  String html = "";
  @override
  void initState() {
    super.initState();

    String metaData =
        widget.metaData == null ? 'null' : jsonEncode(widget.metaData);
    String apiKey = '"${widget.apiKey}"';
    String businessId = '"${widget.businessId}"';
    String email = '"${widget.email}"';
    String phone = '"${widget.phone}"';
    String firstName = '"${widget.firstName}"';
    String lastName = '"${widget.lastName}"';
    String currency = '"${widget.currency}"';
    String amount = widget.amount;

    html = """
    <!DOCTYPE html>
    <html lang="en">

    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, height=device-height, initial-scale=1.0, minimum-scale=1.0">
    </head>

    <body>
      <!-- <button onclick="showPayment()"> Pay with Alatpay </button> -->

    <script src="https://web.alatpay.ng/js/alatpay.js"></script>
    <script>
        let popup = Alatpay.setup({
            apiKey: $apiKey,
            businessId: $businessId,
            email: $email,
            phone: $phone,
            firstName: $firstName,
            lastName: $lastName,
            metaData: $metaData,
            currency: $currency,
            amount: $amount,

            onTransaction: function (response) {
              paymentsuccess(JSON.stringify(response))
            },

            onClose: function () {
              paymentcancel("payment cancel")
            }
        });

        function showPayment() {
            popup.show();
        }
        showPayment()
    </script>
    </body>

    </html>
    """;
  }

  @override
  Widget build(BuildContext context) {
    final media = MediaQuery.of(context).size;
    return Scaffold(
        body: Center(
      // Look here!
      child: WebViewX(
          dartCallBacks: <DartCallback>{
            DartCallback(
              name: 'paymentsuccess',
              callBack: (message) {
                // consoleLog('message success gotten $message');
                dynamic resp = jsonDecode(message);
                // consoleLog('the resp $resp');
                widget.onTransaction(resp);
              },
            ),
            DartCallback(
              name: 'paymentcancel',
              callBack: (message) {
                // consoleLog('message cancel gotten $message');
                if (widget.onClose == null) {
                  Navigator.pop(context);
                } else {
                  widget.onClose!();
                }
              },
            ),
          },
          width: media.width,
          height: media.height,
          initialContent: html,
          initialSourceType: SourceType.html,
          onWebViewCreated: (controller) {
            webviewController = controller;
          }),
    ));
  }
}


example/lib/src/my_app_bar.dart
// ignore_for_file: file_names

import 'package:flutter/material.dart';
import 'package:font_awesome_flutter/font_awesome_flutter.dart';
import 'package:get/route_manager.dart';

import 'constants.dart';



class MyAppBar extends StatelessWidget implements PreferredSizeWidget {
  final String title;
  final double elevation;
  final Color backgroundColor;
  final List<Widget> actions;

  @override
  Size get preferredSize => const Size.fromHeight(55);
  const MyAppBar({
    super.key,
    required this.title,
    required this.elevation,
    required this.actions,
    required this.backgroundColor,
  });
//========================================= FUNCTIONS ============================================\\

//========================================= Navigation ============================================\\
  void _popContext() => Get.back();

  @override
  Widget build(BuildContext context) {
    return AppBar(
      automaticallyImplyLeading: false,
      elevation: elevation,
      backgroundColor: backgroundColor,
      actions: actions,
      title: Row(
        children: [
          InkWell(
            borderRadius: BorderRadius.circular(24),
            onTap: _popContext,
            mouseCursor: SystemMouseCursors.click,
            child: Container(
              width: 40,
              height: 40,
              decoration: ShapeDecoration(
                color: const Color(0xFFFEF8F8),
                shape: RoundedRectangleBorder(
                  side: const BorderSide(
                    width: 0.50,
                    color: Color(0xFFFDEDED),
                  ),
                  borderRadius: BorderRadius.circular(24),
                ),
              ),
              child: Center(
                child: FaIcon(
                  FontAwesomeIcons.circleArrowLeft,
                  color: kAccentColor,
                ),
              ),
            ),
          ),
          kWidthSizedBox,
          Expanded(
            child: Text(
              title,
              style: const TextStyle(
                color: Color(0xFF151515),
                fontSize: 20,
                fontWeight: FontWeight.w700,
                letterSpacing: -0.40,
              ),
            ),
          ),
        ],
      ),
    );
  }
}


example/lib/src/controllers.dart
import 'package:get/get.dart';

import 'constants.dart';

class ApiProcessorController extends GetxController {
  static Future<dynamic> errorState(data) async {
    try {
      if (data.statusCode == 200) {
        return data.body;
      }
      // errorSnack("Something went wrong");
      return;
    } catch (e) {
      // errorSnack("Check your internet and try again");
      return;
    }
  }

  static void errorSnack(msg) {
    Get.showSnackbar(GetSnackBar(
      title: "ERROR",
      message: "$msg",
      backgroundColor: kErrorColor,
      duration: const Duration(seconds: 2),
    ));
  }

  static void successSnack(msg) {
    Get.showSnackbar(GetSnackBar(
      title: "Successful",
      message: "$msg",
      backgroundColor: kSuccessColor,
      duration: const Duration(seconds: 2),
    ));
  }
}


example/lib/main.dart
import 'dart:io';

import 'package:alatpay_flutter/alatpay_flutter.dart';
import 'package:flutter/material.dart';
import 'package:get/route_manager.dart';

import 'screens/success.dart';
import 'src/constants.dart';
import 'src/controllers.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'AlatPay',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: kAccentColor),
        useMaterial3: true,
      ),
      home: const PaymentScreen(),
    );
  }
}

class PaymentScreen extends StatefulWidget {
  const PaymentScreen({Key? key}) : super(key: key);

  @override
  State<PaymentScreen> createState() => _PaymentScreenState();
}

class _PaymentScreenState extends State<PaymentScreen> {
//============= Boolean variables ===============\\
  bool isLoading = false;

//============= Function ===============\\
  void pay() {
    String apiKey = "alatPayPrimaryKey";
    String businessId = "alatPayBuinessId";
    String email = "johndoe@gmail.com";
    String phone = "+2348047589684";
    String firstName = "John";
    String lastName = "Doe";
    String currency = 'NGN';
    String amount = "20000";

    setState(() {
      isLoading = true;
    });

    //=========== Not so important ===========\\

    // Map meta = {
    //   "the_order_id": "order_id",
    //   'client_id': "user_id"
    // };

    //=========== End =============\\
    try {
      Navigator.push(
        context,
        MaterialPageRoute(builder: (context) {
          return AlatPayFlutter(
            apiKey: apiKey,
            businessId: businessId,
            email: email,
            phone: phone,
            firstName: firstName,
            lastName: lastName,
            currency: currency,
            amount: amount,
            // metaData: meta,
            onClose: () {
              Get.close(2);
            },
            onTransaction: (response) async {
              //============ Do something =============\\
              Get.to(const SuccessScreen());
            },
          );
        }),
      );
      setState(() {
        isLoading = false;
      });
    } on SocketException {
      ApiProcessorController.errorSnack("Please connect to the internet");
    } catch (e) {
      consoleLog(e.toString());
    }
    setState(() {
      isLoading = false;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        elevation: 10,
        title: Text(
          "AlatPay",
          style: TextStyle(
            color: Theme.of(context).appBarTheme.backgroundColor,
            fontSize: 20,
            fontWeight: FontWeight.w600,
          ),
        ),
      ),
      body: SafeArea(
          child: Center(
        child: ElevatedButton(
          onPressed: pay,
          style: ElevatedButton.styleFrom(
            disabledBackgroundColor: Colors.grey.shade300,
            backgroundColor: Theme.of(context).primaryColor,
            elevation: 10,
            shape: RoundedRectangleBorder(
              borderRadius: BorderRadius.circular(20),
            ),
            shadowColor: Colors.black.withOpacity(0.4),
          ),
          child: isLoading
              ? const Padding(
                  padding: EdgeInsets.all(10),
                  child: CircularProgressIndicator(color: Colors.white),
                )
              : const Text(
                  "Pay now",
                  textAlign: TextAlign.center,
                  style: TextStyle(
                    color: Colors.white,
                    fontSize: 18,
                    fontFamily: "Sen",
                    fontWeight: FontWeight.w700,
                  ),
                ),
        ),
      )),
    );
  }
}
